File: AndroidArchitecture/mvvm-live/src/main/java/com/codearms/maoqiqi/app/taskdetail/TaskDetailViewModel.java
package com.codearms.maoqiqi.app.taskdetail;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.codearms.maoqiqi.app.Event;
import com.codearms.maoqiqi.app.data.TaskBean;
import com.codearms.maoqiqi.app.data.source.TasksDataSource;
import com.codearms.maoqiqi.app.data.source.TasksRepository;
import com.codearms.maoqiqi.app.utils.MessageMap;

/**
 * Exposes the data to be used in the task detail screen.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/18 17:28
 */
public class TaskDetailViewModel extends ViewModel {

    private String taskId;
    private TasksRepository tasksRepository;

    public final MutableLiveData<String> observableTitle = new MutableLiveData<>();
    public final MutableLiveData<String> observableDescription = new MutableLiveData<>();
    public final MutableLiveData<Boolean> observableCompleted = new MutableLiveData<>();

    private final MutableLiveData<Event<Object>> taskDetailEvent = new MutableLiveData<>();
    private final MutableLiveData<Event<String>> message = new MutableLiveData<>();

    void setTaskId(String taskId) {
        this.taskId = taskId;
    }

    MutableLiveData<Event<Object>> getTaskDetailEvent() {
        return taskDetailEvent;
    }

    MutableLiveData<Event<String>> getMessage() {
        return message;
    }

    public TaskDetailViewModel(TasksRepository tasksRepository) {
        this.tasksRepository = tasksRepository;
    }

    void start() {
        getTask();
    }

    private void getTask() {
        if (isInvalidTaskId()) return;

        tasksRepository.getTask(taskId, new TasksDataSource.GetTaskCallBack() {
            @Override
            public void onTaskLoaded(TaskBean taskBean) {
                observableTitle.setValue(taskBean.getTitle());
                observableDescription.setValue(taskBean.getDescription());
                observableCompleted.setValue(taskBean.isCompleted());
            }

            @Override
            public void onDataNotAvailable() {
                message.setValue(new Event<>(MessageMap.NO_DATA));
            }
        });
    }

    public void completeTask() {
        if (isInvalidTaskId()) return;
        tasksRepository.completeTask(taskId);
        observableCompleted.setValue(true);
        message.setValue(new Event<>(MessageMap.COMPLETE));
    }

    public void activateTask() {
        if (isInvalidTaskId()) return;
        tasksRepository.activateTask(taskId);
        observableCompleted.setValue(false);
        message.setValue(new Event<>(MessageMap.ACTIVE));
    }

    void deleteTask() {
        if (isInvalidTaskId()) return;
        tasksRepository.deleteTask(taskId);
        taskDetailEvent.setValue(new Event<>(new Object()));
    }

    private boolean isInvalidTaskId() {
        if (taskId == null || taskId.equals("")) {
            message.setValue(new Event<>(MessageMap.NO_ID));
            return true;
        }
        return false;
    }
}

File: AndroidArchitecture/mvvm-live-kotlin/src/main/java/com/codearms/maoqiqi/app/taskdetail/TaskDetailViewModel.kt
package com.codearms.maoqiqi.app.taskdetail

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope

import com.codearms.maoqiqi.app.Event
import com.codearms.maoqiqi.app.data.Result
import com.codearms.maoqiqi.app.data.source.TaskRepository
import com.codearms.maoqiqi.app.utils.MessageMap
import kotlinx.coroutines.launch

/**
 * Exposes the data to be used in the task detail screen.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/18 17:28
 */
class TaskDetailViewModel(
    private val tasksRepository: TaskRepository
) : ViewModel() {

    private var taskId: String? = null

    val observableTitle = MutableLiveData<String>()
    val observableDescription = MutableLiveData<String>()
    val observableCompleted = MutableLiveData<Boolean>()

    internal val taskDetailEvent = MutableLiveData<Event<Any>>()
    internal val message = MutableLiveData<Event<String>>()

    internal fun setTaskId(taskId: String?) {
        this.taskId = taskId
    }

    internal fun start() {
        getTask()
    }

    private fun getTask() {
        if (isInvalidTaskId()) return

        viewModelScope.launch {
            val result = tasksRepository.getTask(taskId!!)
            if (result is Result.Success) {
                val taskBean = result.data
                observableTitle.value = taskBean.title
                observableDescription.value = taskBean.description
                observableCompleted.value = taskBean.isCompleted
            } else {
                message.value = Event(MessageMap.NO_DATA)
            }
        }
    }

    fun completeTask() {
        if (isInvalidTaskId()) return
        viewModelScope.launch {
            tasksRepository.completeTask(taskId!!)
            observableCompleted.value = true
            message.value = Event(MessageMap.COMPLETE)
        }
    }

    fun activateTask() {
        if (isInvalidTaskId()) return
        viewModelScope.launch {
            tasksRepository.activateTask(taskId!!)
            observableCompleted.value = false
            message.value = Event(MessageMap.ACTIVE)
        }
    }

    internal fun deleteTask() {
        if (isInvalidTaskId()) return
        viewModelScope.launch {
            tasksRepository.deleteTask(taskId!!)
            taskDetailEvent.value = Event(Any())
        }
    }

    private fun isInvalidTaskId(): Boolean {
        if (taskId == null || taskId == "") {
            message.value = Event(MessageMap.NO_ID)
            return true
        }
        return false
    }
}